server/routes/api/notifications.js

```diff
@@ -0,0 +1,149 @@
+    await user.save();
+
+    // Gmail requires a 200 status acknowledgment to stop sending notifications
+    console.log(
+      `FINISHED processing notification for ${userEmail}, sending res(200)...`
+    );
+    res.status(200).send("OK");
+  } catch (error) {
+    console.log(error);
+    res.status(500).json({ error: "Error saving campaign" });
+  }
+});
+
+const getNewThreadIds = async (gmail, startHistoryId) => {
+  // get list of email ids that have been received since last check
+  const history = await gmail.users.history.list({
+    userId: "me",
+    startHistoryId: startHistoryId,
+    historyTypes: "messageAdded"
+  });
+
+  if (!history.data.history) {
+    console.log("No new messages");
+    return false;
+  }
+  const messageData = history.data.history.flat();
+
+  // get only the threads that have received new messages
+  const messagesAdded = messageData
+    .filter(thread => thread.messagesAdded)
+    .map(thread => thread.messagesAdded);
+
+  if (messagesAdded.length === 0) {
+    console.log("No new messages added");
+    return false;
+  }
+
+  // get just the threadIds of the messages
+  return messagesAdded
+    .map(message => {
+      return message.map(msg => {
+        return msg.message.threadId;
+      });
+    })
+    .flat();
+};
+
+/*
+ * Gets email
+ * @param {string} threadId - id of gmail email thread to read
+ * @param req.query.redirectUrl {string} - URL to redirect to
+ * after Google auth process
+ */
+const getEmail = async (gmailToken, startHistoryId) => {
+  const oAuth2Client = new google.auth.OAuth2(
+    googleClientId,
+    googleClientSecret,
+    googleRedirectUrl
+  );
+  oAuth2Client.setCredentials({ refresh_token: gmailToken });
+  const gmail = google.gmail({ version: "v1", auth: oAuth2Client });
+
+  const threadIds = await getNewThreadIds(gmail, startHistoryId);
+  if (threadIds === false) return false;
+
+  // Find the thread records that match the ids
+  const threadRecords = await Promise.all(
+    threadIds.map(async threadId => {
+      const threadRecord = await Thread.findOne({
+        threadId: threadId
+      });
+      // Check if this is first email in the thread, in which case it was the email mailsender sent
+      if (threadRecord && !threadRecord.receivedPush) {
+        threadRecord.receivedPush = true;
+        await threadRecord.save();
+        return false;
+      }
+      return threadRecord;
+    })
+  );
+
+  // Update the status and replied counters of relevant records
+  await Promise.all(
+    threadRecords.map(async threadRecord => {
+      if (threadRecord) {
+        const campaign = await Campaign.findById(threadRecord.campaignId);
+        const step = await Step.findById(threadRecord.stepId);
+
+        const prospectIndex = campaign.prospects.findIndex(prospect =>
+          prospect.prospectId.equals(threadRecord.prospectId)
+        );
+        campaign.prospects[prospectIndex].status = "Replied";
+        campaign.stepsSummary.replied++;
```

Comment:
Do you think we ever will have duplicated notifications? Just worried that these summary statistics won't be accurate

Reply:
Yeah, I agree it's very dangerous. I guess it would be better to make this a virtual field in Mongoose schema that counts the statuses? Or we just do the count on the back-end when we need it?

Reply:
Leaving this as is for now, given time constraints for pres